WPMedia/arc-themes-components
# TMEDIA-686 - Carousel Component

# UPDATE Wed 3/23:  

@matthewroach expressed more a11y concerns in https://arcpublishing.atlassian.net/browse/TMEDIA-769.  After some research as well as studying what https://accessible360.github.io/accessible-slick/ was doing, I addressed those concerns by applying those needed roles and aria attributes to the DOM structure of react-slick.  I have also updated the custom pager story with a more a11y acceptable implementation. Also the last unit test covers the new functionality to ensure the dom is being updated appropriately.  I feel at this point we have covered a11y concerns and that any further a11y concerns should be conducted through an a11y agency.  While I understand that this adds a bit of patchwork to the solution, lets keep in mind the many other benefits this component provides. Also moving forward, there was an opportunity to investigate a11y concerns during the approval process of https://arcpublishing.atlassian.net/browse/TMEDIA-698, so we need to address these concerns earlier in the process.

# UPDATE Tues 3/22:  

There were some concerns regarding the react-slick component and a11y.  I have ran audits on all stories in both storybook and lighthouse and there are no errors with the component itself.  Certainly, you can break accessibility by choosing colors and button implementation poorly., but that is not the fault of react-slick.  Also, the accessibility property is enabled by default and so if that is explicitly turned off by passing that into the props, a11y could be affected as well.  Finally, I updated the custom prev/next button in the final story to show a more robust accessible demonstration.

# Overview
1. This PR is now ready for review.
2. Unit tests and Storybook stories have been completed.
3. Styles have been formatted per our team discussion this week. Styles have been kept low-fidelity, though some styles have been added to `.storybook/theme.scss` to maintain clarity for Storybook stories.

## Ticket

- [TMEDIA-686](https://arcpublishing.atlassian.net/browse/TMEDIA-686)

## Description

### .editorconfig

We may want to disregard the `.editorconfig` file if it's causing issues or is unneeded. This was added this temporarily to prevent indentation errors in our editors. It will most-likely be removed prior to the final PR, unless wanted or needed.

### React Slick Properties
1. All React Slick properties were brought in as `Carousel` properties to facilitate decision-making on [TMEDIA-699](https://arcpublishing.atlassian.net/browse/TMEDIA-699) and [TMEDIA-700](https://arcpublishing.atlassian.net/browse/TMEDIA-700). It may then be helpful to comment out any unused properties to keep them easily available for upcoming component tickets requiring additional `Carousel` functionality (e.g. using `autoplay` and `autoplaySpeed` for the `Gallery` component).
    
4. The term “Slide” is used in several React Slick properties like `slidesToScroll` and `slidesToShow`. we preserved this term as is, though we may want to consider if “Slide” is the nomenclature to offer. Another possible term to use instead could be “Child Component.”

### Dependencies
This PR adds two dependencies, `react-slick` and `slick-carousel`.  The `react-slick` package contains the `Carousel` underlying React component, but that repository does not have the required Sass files.  There are two Sass files need for the React component, the foundational and the theme.  We included the foundational in the carousel’s `_index.scss` file to keep it in sync in case of any updates from the publisher.  However, we copied in the theme Sass into `_index.scss` for two fold:

1.  To adjust the theme to our needs 
2. To refactor it from using Sass vars to using CSS custom properties to properly tie it into the advanced styling system.  

If we are against linking to the foundational Sass, then we can copy that in, but we then should freeze the version of the `react-slick` component we are currently using.


### Design Patterns

In `/src/components/Carousel/``_index.scss` we are attempting to refactor the slider’s theme to tie into the advanced styling system.  We are doing this by referencing css custom properties that could be referenced in the style’s manifest.  If it’s not defined, then we provide default values.  

This has been done two different ways as seen in the `Paragraph` component and the `Pill` component. We wanted to ensure we were both using a common pattern and citing when inconsistencies are discovered. Any input for this pattern would be appreciated.

#### References:

1. In `Pill`:  https://github.com/WPMedia/arc-themes-components/blob/faa0c4aaef07214fcad6e80d7fae7aea4605faa3/src/components/pill/_index.scss#L4

2. In `Paragraph`: https://github.com/WPMedia/arc-themes-components/blob/faa0c4aaef07214fcad6e80d7fae7aea4605faa3/src/components/paragraph/_index.scss#L4
















##
## Acceptance Criteria

_copy from ticket_

## Test Steps

1. Checkout branch - `git checkout TMEDIA-686`
2. Update dependencies - `npm i`
5. Run Storybook `npm run storybook`
6. Check out the button storybook documentation

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._


## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._
